File: api/views/market.py
import datetime

from django.utils import timezone
from django_ratelimit.decorators import ratelimit
from django.utils.decorators import method_decorator
from rest_framework.response import Response
from rest_framework.views import APIView
from rest_framework import status

from market.models import Asset, Exchange, FXRate, PriceCandle
from trading.models import Order, OrderStatus, Position
from wallets.models import Wallet
from market.services.candles import get_asset_timezone, get_candles_for_range

from api.serializers.market import (
    AssetDetailSerializer,
    AssetListSerializer,
    ExchangeSerializer,
    FxRateSerializer,
)


RANGE_TO_DAYS = {
    "hour": 1 / 24,
    "day": 1,
    "month": 30,
    "6m": 180,
    "year": 365,
}


@method_decorator(ratelimit(key='user', rate='60/m', block=True), name='get')
class ExchangeListView(APIView):
    def get(self, request):
        exchanges = Exchange.objects.prefetch_related('asset_set__currency').all()
        data = []
        for exchange in exchanges:
            assets = [
                a for a in exchange.asset_set.all() if a.is_active
            ]
            serializer = ExchangeSerializer(exchange)
            exchange_data = serializer.data
            exchange_data['assets'] = AssetListSerializer(assets, many=True).data
            exchange_data['asset_count'] = len(assets)
            data.append(exchange_data)

        # Sort: open exchanges first, then by hours until open, then name
        data.sort(key=lambda e: (
            not e['is_open'],
            e['hours_until_open'] if e['hours_until_open'] is not None else 999999,
            e['name'].lower(),
        ))
        return Response(data)


@method_decorator(ratelimit(key='user', rate='60/m', block=True), name='get')
class ExchangeDetailView(APIView):
    def get(self, request, exchange_code):
        try:
            exchange = Exchange.objects.prefetch_related('asset_set__currency').get(code=exchange_code)
        except Exchange.DoesNotExist:
            return Response({'error': 'Exchange not found'}, status=status.HTTP_404_NOT_FOUND)

        assets = [a for a in exchange.asset_set.all() if a.is_active]
        assets.sort(key=lambda a: a.ticker)

        data = ExchangeSerializer(exchange).data
        data['assets'] = AssetListSerializer(assets, many=True).data
        return Response(data)


@method_decorator(ratelimit(key='user', rate='60/m', block=True), name='get')
class AssetDetailView(APIView):
    def get(self, request, exchange_code, ticker):
        try:
            asset = Asset.objects.select_related('exchange', 'currency').get(
                exchange__code=exchange_code,
                ticker=ticker,
            )
        except Asset.DoesNotExist:
            return Response({'error': 'Asset not found'}, status=status.HTTP_404_NOT_FOUND)

        # User wallet for this asset's currency
        wallet = Wallet.objects.filter(
            user_id=request.user.id,
            currency=asset.currency,
        ).first()

        # User position for this asset
        position = Position.objects.filter(
            user_id=request.user.id,
            asset=asset,
        ).first()

        # Pending orders for this asset
        pending_orders = list(
            Order.objects.filter(
                user_id=request.user.id,
                asset=asset,
                status=OrderStatus.PENDING,
            ).order_by('-created_at')[:5]
        )

        serializer = AssetDetailSerializer(asset, context={
            'wallet': wallet,
            'position': position,
            'pending_orders': pending_orders,
        })
        return Response(serializer.data)


@method_decorator(ratelimit(key='user', rate='30/m', block=True), name='get')
class ChartDataView(APIView):
    """Mirrors the existing asset_performance_chart_data_view, now under /api/."""

    def get(self, request, exchange_code, ticker):
        try:
            asset = Asset.objects.select_related('currency').get(
                exchange__code=exchange_code,
                ticker=ticker,
            )
        except Asset.DoesNotExist:
            return Response({'error': 'Asset not found'}, status=status.HTTP_404_NOT_FOUND)

        range_key = request.GET.get('range', 'month')
        tz = get_asset_timezone(asset)
        now_local = timezone.now().astimezone(tz)

        if range_key == 'hour':
            start_at = now_local - datetime.timedelta(hours=1)
            candles = get_candles_for_range(asset, start_at=start_at, end_at=now_local, interval_minutes=5)
            return Response({'chart_type': 'candlestick', 'candlestick_data': candles, 'currency_code': asset.currency.code})

        if range_key == 'day':
            start_at = now_local - datetime.timedelta(days=1)
            candles = get_candles_for_range(asset, start_at=start_at, end_at=now_local, interval_minutes=60)
            return Response({'chart_type': 'candlestick', 'candlestick_data': candles, 'currency_code': asset.currency.code})

        if range_key == 'month':
            start_at = now_local - datetime.timedelta(days=29)
            candles = get_candles_for_range(asset, start_at=start_at, end_at=now_local, interval_minutes=1440)
            return Response({'chart_type': 'candlestick', 'candlestick_data': candles, 'currency_code': asset.currency.code})

        # 6m / year — line chart
        start_days = RANGE_TO_DAYS.get(range_key, RANGE_TO_DAYS['month'])
        start_at = now_local - datetime.timedelta(days=start_days - 1)
        daily_candles = PriceCandle.objects.filter(
            asset=asset,
            interval_minutes=1440,
            start_at__gte=start_at,
            start_at__lte=now_local,
        ).order_by('start_at')

        line_series = [
            {'x': c.start_at.date().isoformat(), 'y': float(c.close_price)}
            for c in daily_candles
        ]
        return Response({'chart_type': 'line', 'line_series': line_series, 'currency_code': asset.currency.code})


@method_decorator(ratelimit(key='user', rate='30/m', block=True), name='get')
class FxRatesView(APIView):
    """All FX rates, used by the wallet FX transfer preview in React."""

    def get(self, request):
        rates = FXRate.objects.select_related('base_currency', 'target_currency').all()
        serializer = FxRateSerializer(rates, many=True)
        return Response(serializer.data)

File: frontend/src/api/market.ts
import api from '@/lib/axios'

export interface Exchange {
  name: string
  code: string
  timezone: string
  open_time: string
  close_time: string
  is_open: boolean
  hours_until_open: number | null
  assets: Asset[]
  asset_count: number
}

export interface Asset {
  ticker: string
  name: string
  asset_type: string
  currency_code: string
  exchange_code: string
  is_active: boolean
  current_price: string | null
}

export interface AssetDetail {
  ticker: string
  name: string
  asset_type: string
  currency_code: string
  exchange_code: string
  exchange_name: string
  exchange_open_time: string
  exchange_close_time: string
  exchange_timezone: string
  is_exchange_open: boolean
  is_active: boolean
  current_price: string | null
  user_wallet: {
    balance: string
    available_balance: string
    pending_balance: string
    currency_code: string
  } | null
  user_position: {
    has_position: boolean
    quantity?: string
    available_quantity?: string
    average_cost?: string
    pending_quantity?: string
  }
  pending_orders: PendingOrder[]
}

export interface PendingOrder {
  id: number
  side: 'BUY' | 'SELL'
  order_type: 'MARKET' | 'LIMIT'
  quantity: string
  limit_price: string | null
  status: string
  created_at: string
}

export interface ChartData {
  chart_type: 'candlestick' | 'line'
  candlestick_data?: { x: number; o: number; h: number; l: number; c: number }[]
  line_series?: { x: string; y: number }[]
  currency_code: string
}

export interface FxRate {
  from_currency: string
  to_currency: string
  rate: string
  last_updated: string
}

export async function getExchanges(): Promise<Exchange[]> {
  const { data } = await api.get('/api/market/exchanges/')
  return data
}

export async function getExchange(code: string): Promise<Exchange> {
  const { data } = await api.get(`/api/market/exchanges/${code}/`)
  return data
}

export async function getAsset(exchangeCode: string, ticker: string): Promise<AssetDetail> {
  const { data } = await api.get(`/api/market/assets/${exchangeCode}/${ticker}/`)
  return data
}

export async function getChartData(
  exchangeCode: string,
  ticker: string,
  range: string,
): Promise<ChartData> {
  const { data } = await api.get(`/api/market/data/${exchangeCode}/${ticker}/`, {
    params: { range },
  })
  return data
}

export async function getFxRates(): Promise<FxRate[]> {
  const { data } = await api.get('/api/wallets/fx-rates/')
  return data
}
